cmd/otel-allocator/allocation/allocator.go

```diff
@@ -88,119 +88,151 @@ func (allocator *Allocator) Collectors() map[string]*collector {
+}
+
+// getCollectorChanges returns the new and removed collectors respectively.
+// This method is called from within SetCollectors, which acquires the needed lock.
+func (allocator *Allocator) getCollectorChanges(collectors []string) ([]string, []string) {
+	var newCollectors []string
+	var removedCollectors []string
+	// Used as a set to check for removed collectors
+	tempCollectorMap := map[string]bool{}
+	for _, s := range collectors {
+		if _, found := allocator.collectors[s]; !found {
+			newCollectors = append(newCollectors, s)
+		}
+		tempCollectorMap[s] = true
+	}
+	for k := range allocator.collectors {
+		if _, found := tempCollectorMap[k]; !found {
+			removedCollectors = append(removedCollectors, k)
+		}
+	}
+	return newCollectors, removedCollectors
+}
+
+// SetTargets accepts a list of targets that will be used to make
 // load balancing decisions. This method should be called when there are
 // new targets discovered or existing targets are shutdown.
-func (allocator *Allocator) SetWaitingTargets(targets []TargetItem) {
-	// Dump old data
+func (allocator *Allocator) SetTargets(targets []TargetItem) {
+	timer := prometheus.NewTimer(timeToAssign.WithLabelValues("SetTargets"))
+	defer timer.ObserveDuration()
+
 	allocator.m.Lock()
 	defer allocator.m.Unlock()
-	allocator.targetsWaiting = make(map[string]TargetItem, len(targets))
-	// Set new data
-	for _, i := range targets {
-		allocator.targetsWaiting[i.hash()] = i
+
+	// Make the temp map for access
+	tempTargetMap := make(map[string]TargetItem, len(targets))
+	for _, target := range targets {
+		tempTargetMap[target.hash()] = target
+	}
+
+	// Check for removals
+	for k, target := range allocator.targetItems {
+		// if the old target is no longer in the new list, remove it
+		if _, ok := tempTargetMap[k]; !ok {
+			allocator.collectors[target.Collector.Name].NumTargets--
+			delete(allocator.targetItems, k)
+			targetsPerCollector.WithLabelValues(target.Collector.Name).Set(float64(allocator.collectors[target.Collector.Name].NumTargets))
+		}
+	}
+
+	// Check for additions
+	for k, target := range tempTargetMap {
+		// Do nothing if the item is already there
+		if _, ok := allocator.targetItems[k]; ok {
+			continue
+		} else {
+			// Assign a collector to the new target
+			allocator.addTargetToTargetItems(&target)
+		}
 	}
 }
 
 // SetCollectors sets the set of collectors with key=collectorName, value=Collector object.
 // This method is called when Collectors are added or removed.
 func (allocator *Allocator) SetCollectors(collectors []string) {
 	log := allocator.log.WithValues("component", "opentelemetry-targetallocator")
+	timer := prometheus.NewTimer(timeToAssign.WithLabelValues("SetCollectors"))
+	defer timer.ObserveDuration()
 
 	allocator.m.Lock()
 	defer allocator.m.Unlock()
+	newCollectors, removedCollectors := allocator.getCollectorChanges(collectors)
 	if len(collectors) == 0 {
 		log.Info("No collector instances present")
 		return
```

Comment:
I think we should move this check above mutex locking - if there are no `collectors`, we shouldn't do any unnecessary work like acquiring a lock or finding the collector changes.

Reply:
What happens here if we scale the collectors to 0? If I'm not mistaken in that case the previously set collectors will still be kept.

Reply:
If we were to scale collectors to 0, there would be nothing querying the target allocator so nothing would observe incorrect information. As soon as that count is changed, we would observe reallocation occur.

Reply:
True, but the `collectorsAllocatable` metric would still show the last number of allocatable collectors

Reply:
ah that's right, I'll update the metric at the beginning of this method.